certora/specs/erc4626.spec

```diff
@@ -0,0 +1,790 @@
+    totalAssets() returns (uint256) envfree
+    pool.getReserveNormalizedIncome(address) returns (uint256)
+    getStaticATokenUnderlying() returns (address) envfree
+    upperBound(uint256) returns (uint256) envfree
+    GetRAY() returns (uint256) envfree
+    ATok._underlyingAsset() returns (address) envfree
+    previewWithdraw(uint256) returns (uint256) envfree
+    previewRedeem(uint256) returns (uint256) envfree
+    maxRedeem(address) returns (uint256) envfree
+    maxDeposit(address) returns (uint256) envfree
+    ATok.allowance(address, address) returns (uint256) envfree
+    previewMint(uint256) returns (uint256) envfree
+    previewWithdraw(address) returns (uint256) envfree
+    maxWithdraw(address) returns (uint256) envfree
+}
+
+
+definition RAY() returns uint256 = 10^27;
+definition HALF_RAY() returns uint256 = 5 * 10^27;
+
+///////////////// PREVIEW RULES ///////////////////////
+
+/***
+* rule to check the following for the previewDeposit function:
+* _1. MUST return as close to and no more than the exact amount of Vault shares that would 
+*     be minted in a deposit call in the same transaction. I.e. deposit should return the same or more shares as previewDeposit if called in the same transaction.
+* 2. Must not account for maxDeposit limit or the allowance of asset tokens
+*  a. Check against maxDeposit
+*  b. Check against allowance
+* 3. Must be inclusive of fees
+* 4. Must not revert due to vault specific user/global limits
+*/
+// STATUS: Verified, that the amount returned by previewDeposit is exactly equal to that returned by the deposit function.
+//         This is a stronger property than the one required by the EIP.
+
+rule previewDepositAmountCheck(){
+    env e;
+    uint256 assets;
+    address receiver;   
+    uint256 previewShares;
+    uint256 shares;
+
+    previewShares = previewDeposit(assets);
+    shares = deposit(e, assets, receiver);
+
+    assert previewShares == shares,"preview shares should be equal to actual shares";
+    // assert assets > maxdeposit() => !lastReverted; //investige this 
+}
+
+
+// checking that the previewDeposit value is independent of allowance of assets
+// STATUS: Verified
+
+rule previewDepositIndependentOfAllowance()
+{
+// allowance of currentContract for asset transfer from msg.sender to   
+    address user;
+    uint256 ATokAllowance1 = ATok.allowance(currentContract, user);
+    uint256 assets1;
+    uint256 assets2;
+    uint256 assets3;
+    require assets1 < ATokAllowance1;
+    uint256 previewShares1 = previewDeposit(assets1);
+
+    env e1;
+    address receiver1;
+    deposit(e1, assets2, receiver1);
+    
+    uint256 ATokAllowance2 = ATok.allowance(currentContract, user);
+    require assets1 == ATokAllowance2;
+    uint256 previewShares2 = previewDeposit(assets1);
+    
+    env e2;
+    address receiver2;
+    deposit(e2, assets3, receiver2); 
+    
+    uint256 ATokAllowance3 = ATok.allowance(currentContract, user);
+    require assets1 > ATokAllowance3;
```

Comment:
The rule may pass vacuously. `asset1 `is not havocked between the requires, so it must obey the conjunction of all the required.